### PR TITLE
Add `warp` based Http JSON RPC node

### PR DESCRIPTION
**Pull request by cryptoAtwill**
This pr adds the JSON RPC node based on `warp`.

# Examples
```rust
use agent::node::config::IPCAgentJsonRPCNodeConfig;
use agent::node::jsonrpc::IPCAgentJsonRPCNode;

#[tokio::main]
async fn main() {
    let n = IPCAgentJsonRPCNode::new(IPCAgentJsonRPCNodeConfig::default());
    n.run().await;
}
```

The following PRs are planned as such:
1. Add in the `IPCAgentJsonRPCHandler` trait that will be shared by all the RPC handlers.
2. Auto routing of `method` and `params` to the Handler.

Btw, my local cargo pipeline will always format the existing code. We should do another PR to add in cicd checks and format code one shot, #10 .

### CODE REVIEW

File: src/jsonrpc/mod.rs
Comment (adlrocha) [inline src/jsonrpc/mod.rs:45]: This `dead_code` directive should not be needed once you rebase from master, right? The `LotusClient` makes use of the `JsonRpcClientImpl` IIRC, right?

File: src/node/config.rs
Comment (adlrocha) [inline src/node/config.rs:6]: We should make these configurable either through the IPC agent config file, or through a CLI flag when we start the IPC agent deamon. I guess it makes more sense to adress it right away in a follow-up PR. Can you add an issue or prepare a PR to tackle this? Thanks!

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:36]:
```suggestion
        log::info!("IPC agent rpc node listening at {:?}", self.config.addr());
```

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:21]: `addr` is an `Option` in the config, are we sure that this works? If I am not mistaken the default for an `Option` is `None`. I don't know the behavior of `run` over a `None` address, but we should probably figure out a better way to pass a default configuration.

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:43]: Can you document what this filter does: 
- Listen to POST requests on the `DEFAULT_JSON_RPC_ENDPOINT` 
- Extract the body of the request. 
- Pass it to to the process function.

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:98]: Can we extract these to a `types.rs` mod?

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:52]: Same here, it would be great to have a better description of what this function does:
- Route the request to the right handler.
- Send request.

Also, what about calling it `handle_request` instead of `process`?

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:63]: Why can't the deserialization into json be part of the filter (the function above), and then this function exclusively handle the routing/handling of functions? 

I would move this logic of deserializing the body into json up to the filter, and I would make this function be the one that picks up the request and receives the request and tailors the response.

You can add a placeholder here or a sample `echo` handler here to showcase how that would be structured.

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:84]: Can we have a bit more of documentation here? Say that this are the standard types for our json-rpc and that follow a structure similar to the one of the Ethereum RPC.

**Review by adlrocha — CHANGES_REQUESTED** [review]
Way more simple than the initial approach, thank you @cryptoAtwill. A few suggestions here and there but we are in the right path.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:98]: reason I put it here because in the scope of the current PR, there is no other struct or module using this. It's completely internal to this file. Maybe I will move it to a separate file when there is a use case for this?

File: src/node/config.rs
Comment (cryptoAtwill) [reply src/node/config.rs:6]: In the configuration, we have a filed to handle the user input. I was thinking the default should be when user has not input anything. If this is from config, then the question is what happens if user does not pass in those default addr or protocol? Then we are asking them to define both the default and the expected url?

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:21]: in the code, it's calling `config.addr()` which handles the `None` and convert to the address to a predefined socket addr.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:52]: ok

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs]: merged

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:63]: The main reason is because this way I dont really have to write another warp middleware, which should terminate the execution if deserialization fails. I was thinking of handling the routing by creating another method in this function body, will that work?

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:45]: let me try

File: src/node/jsonrpc.rs
Comment (adlrocha) [reply src/node/jsonrpc.rs:98]: Gotcha, so they are not publicly exposed yet. I was expecting them to be in the future once the CLI and external libraries need these types to perform jsonrpc requests. I am fine leaving them here for now then, nonetheless can you move them then to the beginning of the file, before the functions declarations so we have first the type and trait definitions and then the functions?

File: src/node/jsonrpc.rs
Comment (adlrocha) [reply src/node/jsonrpc.rs:52]: Please, remember to resolve the threads once the comment has been address (that way I know where to look when I do the second pass to the review) 🙏

File: src/node/jsonrpc.rs
Comment (adlrocha) [reply src/node/jsonrpc.rs:63]: I think is cleaner to have the deserialization on the top filter and the routing in this function. That way we have a clear separation of roles. Why can't you terminate the execution in the execution if the deserialization fails?

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:21]: The other way I considered was actually in serde default deserialization, but that would result in losing the `#[derive(Default)]` and one cannot simply do:
```
IPCAgentJsonRPCNodeConfig{ addr: ..., ..Default::default() }
```
Any other methods you have in mind?

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:84]: updated and resolve this.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:63]: Nah, just need to write more code which probably it's a style preference and might not have liked. I have updated the code accordingly, check it out:)

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:52]: updated and resolving this

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:98]: updated and resolving this.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:43]: updated and resolving this.

File: src/node/config.rs
Comment (adlrocha) [reply src/node/config.rs:6]: In the design doc we considered that the daemon always accepts the path of a config file (or use the default config location) as input: https://docs.google.com/document/d/14lkRRv6MQYnuEfp2GoGngdD8Q5YgfE38D8HTZWKgKf4/edit#heading=h.i71qsbf2iwn5

With this in mind I don't think the default address will ever be not set, as a config file should be picked up, even if it is the default one. In any case, I guess we can address this in the coming PR with the daemon command and discuss how to handle it there. WDYT?

File: src/node/jsonrpc.rs
Comment (adlrocha) [reply src/node/jsonrpc.rs:41]: Oh, I see... so `addr` is private and when accessing it through `addr()` it auto-assigns the default value. The thing that I am now wondering, is if once we force the use of a config file as input we need `addr` to be an `Option` at all, but I guess we can discuss that in the next iteration.

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:66]:
```suggestion
/// - Pass it to to the json_rpc_filter to deserialize into a jsonrpc request.
```

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:83]:
```suggestion
/// Main function responsible for handling and routing jsonrpc requests to the right underlying handler according to the method
```

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:75]:
```suggestion
// Filter that deserializes the body of the request into a jsonrpc request.
```

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:143]: Is this command correct?

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:162]: Same here

File: src/node/jsonrpc.rs
Comment (adlrocha) [inline src/node/jsonrpc.rs:159]: If you see that it is not really hard, would you mind adding some quick tests with the error cases for `handle_rejection`? Just in case we introduce regressions in the future. Thanks!

**Review by adlrocha — APPROVED** [review]
A new round of light suggestions, but overall I think it is in a good state now. Thanks. Summoning @hmoniz in case he wants to do a quick pass before merging.

File: src/node/config.rs
Comment (cryptoAtwill) [reply src/node/config.rs:6]: Yeah, I can use that config file. But in the file you sent, I dont see a config parameter for the addr of the current node? Or it's not added yet?

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:41]: Yep, make sense. If we are using the config file, I dont even need this config struct. I can use that config file. But just need to define a field for this addr.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:159]: yep, should be quick. Will add it.

File: src/node/jsonrpc.rs
Comment (hmoniz) [inline src/node/jsonrpc.rs:124]: nit: the convention is to the plural form, i.e., `mod tests`.

File: src/node/jsonrpc.rs
Comment (hmoniz) [inline src/node/jsonrpc.rs:23]: This must also include an `error` field to conform with the JSON-RPC spec (see https://www.jsonrpc.org/specification#response_object). The serialization then uses one of `result` and `error` depending on the outcome of the processing of the request.

**Review by hmoniz — CHANGES_REQUESTED** [review]
This PR is definitely on a good trajectory. I left a few comments. The most important ones are related to the config and the JSON-RPC response missing the error field.

I also suggest a bit of a code restructuring, which perhaps is better addressed in a follow-up PR. The suggestion is the following: Let's have a `jsonrpc` module with submodules `client` and `server` (or `node` - I have a slight preference for `server` but curious to hear others thoughts).

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:23]: right right, then in this case I think we can do something like this:
```
struct JSONRPCResponse<T: Serialize> {
    pub id: u16,
    pub jsonrpc: String,
    #[some serde config to ignore serialization if None]
    pub result: Option<T>,
    #[some serde config to ignore serialization if None]
    pub error: Option<String>, // we can use T is we want support struct.
}
```
Then, we can based to `Result` to set `result` or `error`.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:13]: Yep yep, I will do this in a follow up PR, seems it can be shared and maybe placed in to `util` or `common` module or we can just import from `jsonrpc` mod?

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:202]: yep yep, quite strange though, I don't see the changes locally. Let me check how did the formatting slipped in.

File: src/node/config.rs
Comment (adlrocha) [reply src/node/config.rs:6]: Yes, that's a good point, @hmoniz. Let's call it `server` or if we want to be verbose (and just in case we end up with more future servers) we can use something like `AGENT_RPC_SERVER`

File: src/node/jsonrpc.rs
Comment (adlrocha) [reply src/node/jsonrpc.rs:13]: This goes in the line of a previous comment I made, if it is not a lot of work let's just move it right away instead of waiting for a follow-up PR. WDYT?

**Review by adlrocha — COMMENTED** [review]
Thank you folks for the review and addressing the comments. @cryptoAtwill, let's resolve the few pending threads, fix merge conflicts and pipeline it for merge :)

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:143]: oops, you got me. Copy pasted sample from their official doc. Will remove.

File: src/node/config.rs
Comment (cryptoAtwill) [reply src/node/config.rs:6]: Here is what I'm proposing:
Update existing `config.rs` into a mod, then the `Config` struct will stay in the `mod.rs`. The overall folder struct is something like:
```
config/
-- mod.rs
-- subnet.rs  // contains subnet related configs
-- server.rs  // contains server related configs
-- utils.rs  // contains all of the `deserialize_*` functions that can be shared.
```
This way maybe we can add more configs without problem. The other way which I was thinking of is actually putting the configs into each module, i.e. `server.rs` is placed in the module together with rpc node, so that all the code related to rpc node is contained in one folder. Both should be quick changes, I guess it's just a preference, maybe we try putting the configs in one folder first? 

@hmoniz @adlrocha any thoughts?

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:13]: The current `jsonrpc` mod contains the request and response, which was not merged when this code was written. Since those are merged, I can just use those directly.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs:23]: Update: 2023/02/20:
I think it's better not to mix the result and error. I read the [json rpc spec](https://www.jsonrpc.org/specification#response_object) of json_rpc, seems the error can be more complex. It's neater to use another struct just to handle the error.

File: src/node/config.rs
Comment (hmoniz) [reply src/node/config.rs]: I really like your first suggestion as it keeps all the config (1) self-contained and (2) neatly modularized.

Comment (adlrocha) [reply src/node/config.rs]: I like that approach of having a single config crate with topical files. I'll review the linked PR also. 🙏

File: src/node/jsonrpc.rs
Comment (adlrocha) [reply src/node/jsonrpc.rs]: I am fine either way. Whatever works best for you. Whenever this happens feel free to stack a new PR with the proposed change and link it from here so we can have the overview of the future changes and so you are not blocked by this being merged as you did with the config change

Comment (hmoniz) [reply src/node/jsonrpc.rs]: +1 for doing this in a separate PR

File: src/node/jsonrpc.rs
Comment (hmoniz) [reply src/node/jsonrpc.rs]: You can simply have here a field like: 
```
pub error: Option<JsonRpcError>
```
Is this what you're suggesting?

Comment (cryptoAtwill) [reply src/node/jsonrpc.rs]: Refer to this [struct](https://github.com/consensus-shipyard/ipc-agent/blob/1187b226e389d5323c8edc3f02f506534f3a683c/src/server/response.rs#L42). The usage is [here](https://github.com/consensus-shipyard/ipc-agent/blob/1187b226e389d5323c8edc3f02f506534f3a683c/src/server/jsonrpc.rs#L76)

File: src/node/jsonrpc.rs
Comment (hmoniz) [reply src/node/jsonrpc.rs]: BTW, it might be a good idea to address that PR first and then come back to this one.

Comment (cryptoAtwill) [reply src/node/jsonrpc.rs]: I think if it's not a big issue, maybe get this one merged first? Too many stacking up PRs will create quite some conflicts.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs]: As discussed, will close this one first then sharing the request and responses. Resolving first.

File: src/node/jsonrpc.rs
Comment (cryptoAtwill) [reply src/node/jsonrpc.rs]: I have create this issue to track it: #24, resolving this for now.

**Comment by cryptoAtwill** [comment]
Hi @hmoniz @adlrocha, please help take a look since the config PR is merged. 

There is one outstanding issue, which is the shared Json RPC response and request. Based on previous discussion, I will do it in a follow up PR which is tracked in #24.

I think there are no other pending issues (maybe I missed something? lol). Please help take a look, thank you very much!

File: src/config/mod.rs
Comment (hmoniz) [inline src/config/mod.rs:26]: nit: As mentioned in the other PR, I'd rather see here a `HashMap<String, Subnet>` instead of this custom type.

File: src/config/mod.rs
Comment (hmoniz) [inline src/config/mod.rs:66]: I think it would be preferable to have this function and `check_subnets_config` as `#[test]` instead of `read_config`. This way, if there are issues with a specific config section, we know immediately which one is it by looking at the test that is failing.

**Review by hmoniz — APPROVED** [review]
Great job here. LGTM. I just left a couple of suggestions that don't need another review round.

File: src/config/mod.rs
Comment (cryptoAtwill) [reply src/config/mod.rs:26]: my bad, I didn't see this. Will revert this to `HashMap`.

File: src/config/mod.rs
Comment (adlrocha) [reply src/config/mod.rs:67]: Agree! Let's address this one and we should be good to go :pray:

**Review by adlrocha — APPROVED** [review]

File: src/config/mod.rs
Comment (cryptoAtwill) [reply src/config/mod.rs]: Cool, this is updated! Thx @adlrocha for the tip. Resolving this.

**Pull request merged**
